engine/table/src/main/java/io/deephaven/engine/table/impl/locations/local/URIStreamKeyValuePartitionLayout.java

```diff
@@ -0,0 +1,143 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.net.URI;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * Extracts a key-value partitioned table layout from a stream of URIs.
+ */
+public abstract class URIStreamKeyValuePartitionLayout<TLK extends TableLocationKey>
+        extends KeyValuePartitionLayout<TLK, URI> {
+
+    protected final URI tableRootDirectory;
+    private final Supplier<KeyValuePartitionLayout.LocationTableBuilder> locationTableBuilderFactory;
+    private final int maxPartitioningLevels;
+    private final String fileSeparator;
+
+    /**
+     * @param tableRootDirectory The directory to traverse from
+     * @param locationTableBuilderFactory Factory for {@link LocationTableBuilder builders} used to organize partition
+     *        information; as builders are typically stateful, a new builder is created each time this
+     *        {@link KeyValuePartitionLayout} is used to {@link #findKeys(Consumer) find keys}
+     * @param keyFactory Factory function used to generate table location keys from target files and partition values
+     * @param maxPartitioningLevels Maximum partitioning levels to traverse. Must be {@code >= 0}. {@code 0} means only
+     *        look at files in {@code tableRootDirectory} and find no partitions.
+     */
+    public URIStreamKeyValuePartitionLayout(
+            @NotNull final URI tableRootDirectory,
+            @NotNull final Supplier<KeyValuePartitionLayout.LocationTableBuilder> locationTableBuilderFactory,
+            @NotNull final BiFunction<URI, Map<String, Comparable<?>>, TLK> keyFactory,
+            final int maxPartitioningLevels,
+            @NotNull final String fileSeparator) {
+        super(keyFactory);
+        this.tableRootDirectory = tableRootDirectory;
+        this.locationTableBuilderFactory = locationTableBuilderFactory;
+        this.maxPartitioningLevels = Require.geqZero(maxPartitioningLevels, "maxPartitioningLevels");
+        this.fileSeparator = fileSeparator;
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + '[' + tableRootDirectory + ']';
+    }
+
+    /**
+     * Find the keys in the given URI stream and notify the observer.
+     */
+    protected final void findKeys(@NotNull final Stream<URI> uriStream,
+            @NotNull final Consumer<TLK> locationKeyObserver) {
+        final KeyValuePartitionLayout.LocationTableBuilder locationTableBuilder = locationTableBuilderFactory.get();
+        final Deque<URI> targetURIs = new ArrayDeque<>();
+        final Set<String> takenNames = new HashSet<>();
```

Comment:
Either delete this, or revisit when we determine how we want name-duplication-testing to work.